### ISSUE TITLE
NLOSS is positive

### ISSUE THREAD

**Issue by Jeffrey-Sardina** [hugochan/IDGL#18]
I have been running into an odd case during training: sometimes NLOSS takes a positive value.

**My current understanding is this:**
I am targeting downstream classification, so the loss function being used is NLL (negative log likelihood). My understanding is that log likelihood is the log of a values on the interval [0:1]. This means that log must be negative, or 0. Thus, its negative must be positive, or 0.

NLOSS in IDGL is the negative of the NLL loss function, ie -NLL. Thus, it just recovers the original log-likelihood, which is strictly non-positive.

**My observations:**
I have observed cases where it is positive. Oddly, in most of these cases, I will see that, for some loss value L, the accuracies associating with training models that have losses near L or near -L are nearly identical.

This prompted me to think that there is some edge case where a negative is not applied (or is applied once too many times)

One sample of a case where the NLOSS is positive is this:
```
{
    "data_type": "network",
    "dataset_name": "cora",
    "data_dir": "../data/cora/",
    "pretrained": null,
    "task_type": "classification",
    "out_dir": "/workspace/output/models/TwigJob_n5_bdg250_loglr_repeat_fixout/3_0_7/0.7623429221128561",
    "seed": 77,
    "model_name": "GraphClf",
    "scalable_run": true,
    "num_anchors": 7967,
    "hidden_size": 43,
    "use_bert": false,
    "dropout": 0.6635843579072174,
    "feat_adj_dropout": 0.06838793351721184,
    "gl_dropout": 0.548838069426093,
    "bignn": false,
    "graph_module": "gcn",
    "graph_type": "dynamic",
    "graph_learn": true,
    "graph_metric_type": "weighted_cosine",
    "graph_skip_conn": 0.7949061507223335,
    "update_adj_ratio": 0.632239148921817,
    "graph_include_self": false,
    "graph_learn_regularization": true,
    "smoothness_ratio": 0.530333708979554,
    "degree_ratio": 0.804512963814122,
    "sparsity_ratio": 0.10112626099490862,
    "graph_learn_ratio": 0,
    "graph_learn_hidden_size": 33,
    "graph_learn_epsilon": 0,
    "graph_learn_topk": null,
    "graph_learn_num_pers": 6,
    "graph_hops": 4,
    "gat_nhead": 8,
    "gat_alpha": 0.2,
    "optimizer": "adam",
    "learning_rate": 1.634816494138241e-05,
    "weight_decay": 0.002369427396679898,
    "lr_patience": 7,
    "lr_reduce_factor": 0.20687033919483733,
    "grad_clipping": null,
    "grad_accumulated_steps": 6,
    "eary_stop_metric": "nloss",
    "pretrain_epoch": 0,
    "max_iter": 10,
    "eps_adj": 8.5e-05,
    "rl_ratio": 0,
    "rl_ratio_power": 1,
    "rl_start_epoch": 1,
    "max_rl_ratio": 0.99,
    "rl_reward_metric": "acc",
    "rl_wmd_ratio": 0,
    "random_seed": 77,
    "shuffle": true,
    "max_epochs": 27.777777777777775,
    "patience": 100,
    "verbose": 20,
    "print_every_epochs": 500,
    "out_predictions": false,
    "out_raw_learned_adj_path": "cora_idgl_node_anchor_adj.npy",
    "save_params": true,
    "logging": true,
    "no_cuda": false,
    "cuda_id": 0
}
```

The output of this is:
```
NLOSS = 6.22605
ACC = 0.19400
```

Do you have any idea why this would occur? Or is my understanding of how loss is reported incorrect?

Thank you in advance!

Sincerely,
--Jeffrey Sardina

**Comment by hugochan** [comment]
> data

@Jeffrey-Sardina Thank you for your interest to our work! You are right about the NLL loss. However, the overall loss adds up both the NLL loss and graph regularization loss (which can be negative), as shown [here](https://github.com/hugochan/IDGL/blob/master/src/core/model_handler.py#L712). Therefore, the overall loss can be negative, and `nloss` can be positive. Please refer to Eq. (11) in the [paper](https://arxiv.org/pdf/2006.13009.pdf) for more details. 

Back to your run, because your `degree_ratio` value was large, so the connectivity loss (first item in Eq. (11)) became more dominant, which led to negative overall loss.

**Comment by Jeffrey-Sardina** [comment]
Thank you very much! That makes a lot of sense now
